hazelcast/src/main/java/com/hazelcast/query/impl/getters/AbstractJsonGetter.java

```diff
@@ -0,0 +1,196 @@
+        } else if (value.isNull()) {
+            return null;
+        } else if (value.isString()) {
+            return value.asString();
+        }
+        throw new HazelcastSerializationException("Unknown Json type: " + value);
+    }
+
+    protected String[] getPath(String attributePath) {
+        return attributePath.split(DELIMITER);
+    }
+
+    @Override
+    Object getValue(Object obj) throws Exception {
+        throw new HazelcastException("Path agnostic value extraction is not supported");
+
+    }
+
+    @Override
+    Object getValue(Object obj, String attributePath) throws Exception {
+        String[] paths = getPath(attributePath);
+        JsonParser parser = createParser(obj);
+
+        parser.nextToken();
+        for (int i = 0; i < paths.length; i++) {
+            String path = paths[i];
+            String arrayIndexText = getIndexTextOrNull(path);
+            if (arrayIndexText == null) {
+                // non array case
+
+
+                if (!findAttribute(parser, path)) {
+                    return null;
+                }
+            } else {
+                // array case
+                if ("any".equals(arrayIndexText)) {
+                    String lastPath = i + 1 < paths.length ? paths[i + 1] : null;
+                    return getMultiValue(parser, lastPath);
+                } else {
+                    JsonToken token = parser.getCurrentToken();
+                    if (token != JsonToken.START_ARRAY) {
+                        return null;
+                    }
+                    token = parser.nextToken();
+                    int arrayIndex = Integer.parseInt(arrayIndexText);
+                    for (int j = 0; j < arrayIndex; j++) {
+                        if (token == JsonToken.END_ARRAY) {
+                            return null;
+                        }
+                        parser.skipChildren();
+                        token = parser.nextToken();
+                    }
+                }
+            }
+        }
+        Object ret = convertJsonTokenToValue(parser);
+        parser.close();
+        return ret;
+    }
+
+    abstract JsonParser createParser(Object obj) throws IOException;
+
+    private boolean findAttribute(JsonParser parser, String path) throws IOException {
+        JsonToken token = parser.getCurrentToken();
+        if (token != JsonToken.START_OBJECT) {
+            return false;
+        }
+        parser.getCurrentToken();
+        while (true) {
+            token = parser.nextToken();
+            if (token == JsonToken.END_OBJECT) {
+                return false;
+            }
+            if (path.equals(parser.getCurrentName())) {
+                parser.nextToken();
+                return true;
+            } else {
+                parser.nextToken();
+                parser.skipChildren();
+            }
+        }
+    }
+
+    private MultiResult getMultiValue(JsonParser parser, String lastPath) throws IOException {
```

Comment:
Can you add the javadoc telling what this and the other private methods are doing? For instance, I guess this one is traversing an array and adding the scalar values or attributes of an object. It also skips elements which are nested arrays. Is this right?